lib/semantics/check-declarations.cc

```diff
@@ -106,6 +107,41 @@ void CheckHelper::Check(const Symbol &symbol) {
   if (isAssociated) {
     return;  // only care about checking VOLATILE on associated symbols
   }
+  if (symbol.has<ProcBindingDetails>()) {
+    CheckBinding(symbol);
+    return;
+  }
+  if (const auto *details{symbol.detailsIf<DerivedTypeDetails>()}) {
+    CHECK(symbol.scope());
+    CHECK(symbol.scope()->symbol() == &symbol);
+    CHECK(symbol.scope()->IsDerivedType());
+    if (symbol.attrs().test(Attr::ABSTRACT) &&
+        (symbol.attrs().test(Attr::BIND_C) ||
+            (details && details->sequence()))) {
+      messages_.Say("An ABSTRACT derived type must be extensible"_err_en_US);
+    }
+    if (const DeclTypeSpec * parent{FindParentTypeSpec(symbol)}) {
+      const DerivedTypeSpec *parentDerived{parent->AsDerived()};
+      if (!IsExtensibleType(parentDerived)) {
+        messages_.Say("The parent type is not extensible"_err_en_US);
+      }
+      if (!symbol.attrs().test(Attr::ABSTRACT) && parentDerived &&
+          parentDerived->typeSymbol().attrs().test(Attr::ABSTRACT)) {
+        ScopeComponentIterator components{*parentDerived};
+        for (const Symbol &component : components) {
+          if (component.attrs().test(Attr::DEFERRED)) {
+            if (symbol.scope()->find(component.name()) ==
+                symbol.scope()->end()) {
```

Comment:
This code is only looking in the current derived type `symbol` for a binding, but I think there may be issue here if the deferred was actually defined in an intermediate derived type. e.g:
```
module m0

  type, abstract :: t0
  contains
    procedure(swap), deferred, nopass :: q
  end type

  type, abstract, extends(t0) :: t1
  contains
    procedure, nopass :: q => swap
  end type

  ! Error raised here
  type, extends(t1) :: t2
  end type

contains
  subroutine swap(i, j)
    integer :: i, j, tmp
    tmp = i
    i = j
    j = tmp
  end subroutine

end module
```
f18 raises an error:
`Non-ABSTRACT extension of ABSTRACT derived type 't1' lacks a binding for DEFERRED procedure 'q'`
I think it should not since a binding for `q` was defined in `t1`.

Reply:
Good point, thanks.